### ISSUE TITLE
Lint top-level code in test files?

### ISSUE THREAD

**Issue by IndrajeetPatil** [r-lib/lintr#2118]
The _R Packages_ [book](https://r-pkgs.org/testing-design.html#self-sufficient-tests) discusses how top-level code in test files can compromise the self-sufficiency of tests, contaminate the global environment for other test files, and also make the tests more difficult to read and understand.

In particular, it advises:

> The `test-*.R` files below `tests/testthat/` should consist almost entirely of calls to `test_that()`.
> Any other top-level code is suspicious and should be carefully considered for relocation into calls to `test_that()` or to other files that get special treatment inside an R package or from testthat.

So I am wondering if a new linter is in order to detect such code?

---

## Example that will produce lints

I am not completely sure if the skips should be linted here because sometimes you do want to skip an entire test file and a top-level skip would come in handy there.

```r
dat <- data.frame(x = c("a", "b", "c"), y = c(1, 2, 3)) # lint

skip_if(today_is_a_monday()) # lint?

test_that("foofy() does this", {
  expect_equal(foofy(dat), ...)
})

dat2 <- data.frame(x = c("x", "y", "z"), y = c(4, 5, 6)) # lint

skip_on_os("windows") # lint?

test_that("foofy2() does that", {
  expect_snapshot(foofy2(dat, dat2))
})
```

## Example that won't produce lints

```r
test_that("foofy() does this", {
  skip_if(today_is_a_monday())
  
  dat <- data.frame(x = c("a", "b", "c"), y = c(1, 2, 3))
  
  expect_equal(foofy(dat), ...)
})

test_that("foofy() does that", {
  skip_if(today_is_a_monday())
  skip_on_os("windows")
  
  dat <- data.frame(x = c("a", "b", "c"), y = c(1, 2, 3))
  dat2 <- data.frame(x = c("x", "y", "z"), y = c(4, 5, 6))
  
  expect_snapshot(foofy(dat, dat2))
})
```

**Comment by MichaelChirico** [comment]
> contaminate the global environment for other test files

that's not true afaik. each test file is run in a fresh environment unless some global assignment is done.

**Comment by MichaelChirico** [comment]
> unless some global assignment is done.

I'd count library under here 😅

the files aren't run in subprocesses so things like library that have global side effects will survive.

definitely agree calling library() in test* files is bad practice.
